addons/controllers/kapp-controller/kappcontrollerconfig_controller.go

```diff
@@ -0,0 +1,219 @@
+
+			if err := patchHelper.Patch(ctx, kappControllerConfig); err != nil {
+				log.Error(err, "Error patching kappControllerConfig")
+				retErr = err
+			}
+		}
+	}()
+
+	if err := r.ReconcileKappControllerConfigNormal(ctx, kappControllerConfig, cluster, log, &patchConfig); err != nil {
+		log.Error(err, "Error reconciling kappControllerConfig")
+		return ctrl.Result{}, err
+	}
+
+	log.Info("Successfully reconciled kappControllerConfig")
+
+	return ctrl.Result{}, nil
+}
+
+// ReconcileKappControllerConfigNormal reconciles KappControllerConfig CR
+func (r *KappControllerConfigReconciler) ReconcileKappControllerConfigNormal(
+	ctx context.Context,
+	kappControllerConfig *runv1alpha3.KappControllerConfig,
+	cluster *clusterapiv1beta1.Cluster,
+	log logr.Logger,
+	patchConfig *bool) (retErr error) {
+
+	// add owner reference to kappControllerConfig
+	ownerReference := metav1.OwnerReference{
+		APIVersion:         clusterapiv1beta1.GroupVersion.String(),
+		Kind:               cluster.Kind,
+		Name:               cluster.Name,
+		UID:                cluster.UID,
+		BlockOwnerDeletion: pointer.BoolPtr(false),
+	}
+
+	if !clusterapiutil.HasOwnerRef(kappControllerConfig.OwnerReferences, ownerReference) {
+		log.Info("Adding owner reference to kappControllerConfig")
+		kappControllerConfig.OwnerReferences = clusterapiutil.EnsureOwnerRef(kappControllerConfig.OwnerReferences, ownerReference)
+		*patchConfig = true
+	}
+
+	if err := r.ReconcileKappControllerConfigDataValue(ctx, kappControllerConfig, cluster, log); err != nil {
+		log.Error(err, "Error creating kappControllerConfig data value secret")
+		return err
+	}
+
+	// update status.secretRef
+	dataValueSecretName := util.GenerateDataValueSecretName(kappControllerConfig.Name, KappControllerAddonName)
+	if kappControllerConfig.Status.SecretRef != dataValueSecretName {
+		kappControllerConfig.Status.SecretRef = dataValueSecretName
+		*patchConfig = true
+	}
+
+	return nil
+}
+
+// ReconcileKappControllerConfigDataValue reconciles KappControllerConfig data values secret
+func (r *KappControllerConfigReconciler) ReconcileKappControllerConfigDataValue(
+	ctx context.Context,
+	kappControllerConfig *runv1alpha3.KappControllerConfig,
+	cluster *clusterapiv1beta1.Cluster,
+	log logr.Logger) (retErr error) {
+
+	dataValuesSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      util.GenerateDataValueSecretName(kappControllerConfig.Name, KappControllerAddonName),
+			Namespace: cluster.Namespace,
+		},
+	}
+
+	dataValuesSecretMutateFn := func() error {
+		dataValuesSecret.Type = corev1.SecretTypeOpaque
+		dataValuesSecret.Data = map[string][]byte{}
+
+		// marshall the yaml contents
+		kappConfig, err := mapKappControllerConfigSpec(cluster, kappControllerConfig)
+		if err != nil {
+			return err
+		}
+
+		yamlBytes, err := yaml.Marshal(kappConfig)
+		if err != nil {
+			return err
+		}
+
+		dataValueBytes := append([]byte(constants.TKGDataValueFormatString), yamlBytes...)
```

Comment:
Do we still need to do this? I think for packages in general we only want to support a key:value map.

Reply:
We still need this if we plan to add overlay support with `ext.packaging.carvel.dev/ytt-data-values-overlays: {}`. Maybe let's remove for now?

Reply:
yeah lets remove for now